lib/CGameState.cpp

```diff
@@ -3532,8 +3537,16 @@ bool CPathfinder::goodForLandSeaTransition()
 
 CPathfinder::CPathfinder(CPathsInfo &_out, CGameState *_gs, const CGHeroInstance *_hero) : CGameInfoCallback(_gs, boost::optional<PlayerColor>()), out(_out), hero(_hero), FoW(getPlayerTeam(hero->tempOwner)->fogOfWarMap)
 {
-	useSubterraneanGates = true;
+	assert(hero);
+	assert(hero == getHero(hero->id));
+
 	allowEmbarkAndDisembark = true;
+	allowTeleportTwoWay = true;
+	allowTeleportOneWay = true;
+	allowTeleportOneWayRandom = false;
+	allowTeleportWhirlpool = false;
+	if (CGWhirlpool::isProtected(hero))
```

Comment:
CPathfinder object is created on each call of calculatePaths so it's wouldn't change.
How good is that fact above it's different question.